code/autoannotate.py

```diff
@@ -0,0 +1,321 @@
+        that are used in that reference taxonomy file.
+        Default = "s__"
+
+    '''
+
+
+def main(argv):
+    source_fp = ''
+    destination_dir = ''
+    ref_taxa_fp = ''
+    sep = ';'
+    gen = 'g__'
+    sp = 's__'
+
+    try:
+        opts, args = getopt.getopt(argv, "hi:o:r:p:g:s:", ["help",
+                                                           "infile=",
+                                                           "outdir=",
+                                                           "ref_taxa=",
+                                                           "separator",
+                                                           "genus",
+                                                           "species"])
+    except getopt.GetoptError:
+        print(usage)
+        sys.exit(2)
+
+    for opt, arg in opts:
+        if opt in ("-h", "--help"):
+            print(usage)
+            sys.exit()
+        elif opt in ("-i", "--infile"):
+            source_fp = arg
+        elif opt in ("-o", "--outdir"):
+            destination_dir = arg
+        elif opt in ("-r", "--ref_taxa"):
+            ref_taxa_fp = arg
+        elif opt in ("-p", "--separator"):
+            sep = arg
+        elif opt in ("-g", "--genus"):
+            gen = arg
+        elif opt in ("-s", "--species"):
+            sp = arg
+
+    # generate dict of {name: (genus, species, abundances)}
+    with open(source_fp, "r") as source:
+        sample_list = source.readline().strip().split('\t')[1:]
+        taxa = {}
+        for l in source:
+            # convert abundances to float
+            abundances = list(map(float, l.strip().split('\t')[1:]))
+            name = l.strip().split('\t')[0]
+            # level labels (e.g., Silva's 'D_11__') can confabulate this.
+            # Hence, do split on '__' instead of sep to remove level labels
+            taxon = re.split(' |_', name.split(';')[-1])[0:2]
+            if name not in taxa.keys():
+                taxa[name] = (taxon, abundances)
+            else:
+                # if species is replicated, collapse abundances
+                taxa[name] = (taxon, add_lists(taxa[name][1], abundances))
+
+    # parse ref taxonomy
+    with open(ref_taxa_fp, "r") as ref:
+        ref_taxa = {l.strip().split('\t')[1]: (
+                        l.strip().split('\t')[1].split(sep)[-2],
+                        l.strip().split('\t')[1].split(sep)[-1],
+                        l.strip().split('\t')[0]
+                        ) for l in ref}
+
+    # find matching taxonomies
+    species_match = 0
+    genus_match = 0
+    family_match = 0
+    no_match = 0
+    count = len(taxa)
+
+    duplicates = []
+    seq_ids = dict()
+    new_taxa = dict()
+
+    for name, t in taxa.items():
+        species_set = set()
+        genus_set = set()
+        match = 'None'
+
+        # search for match at genus, then species level
+        for full, partial in ref_taxa.items():
+            if t[0][0] in partial[0]:
```

Comment:
This exact matching is really strict - I wonder if you want to use fuzzy matching instead, to help you identify slight differences in spelling or spelling errors. [Fuzzywuzzy](https://github.com/seatgeek/fuzzywuzzy) is one package that does this. This isn't essential, but is a "nice to have" feature that would probably save you some time.